lib/rules/no-unused-vars.js

```diff
@@ -610,6 +717,10 @@ module.exports = {
                             config.destructuredArrayIgnorePattern &&
                             config.destructuredArrayIgnorePattern.test(def.name.name)
                         ) {
+                            if (config.reportUsedIgnorePattern && isUsedVariable(variable)) {
+                                usedIgnoredVars.push(variable);
+                            }
```

Comment:
Instead of collecting these vars and later figuring out why they were considered as used ignored, we could just report them right away with the data we have at this point:

```suggestion
                            if (config.reportUsedIgnorePattern && isUsedVariable(variable)) {
                                context.report({
                                    node: def.name,
                                    messageId: "usedIgnoredVar",
                                    data: {
                                        varName: variable.name,
                                        additional: `. Used elements of array destructuring must not match ${config.destructuredArrayIgnorePattern}`
                                    }
                                });
                            }
```

(the same for catch arguments, params and variables).

That would fix https://github.com/eslint/eslint/pull/17662#issuecomment-1910344522 and https://github.com/eslint/eslint/pull/17662#issuecomment-1959062359.